chain/types/tipset_key.go

```diff
@@ -0,0 +1,132 @@
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+)
+
+// A TipSetKey is an immutable collection of CIDs forming a unique key for a tipset.
+// The CIDs are assumed to be distinct and in canonical order. Two keys with the same
+// CIDs in a different order are not considered equal.
+// TipSetKey is a lightweight value type, and may be compared for equality with ==.
+type TipSetKey struct {
+	// The internal representation is a concatenation of the bytes of the CIDs, each preceded by
+	// uint32 specifying the length of the CIDs bytes, and the whole preceded by a uint32
+	// giving the number of CIDs.
+	// These gymnastics make the a TipSetKey usable as a map key.
+	// This representation is slightly larger than strictly necessary: CIDs do carry a prefix
+	// including their length, but the cid package doesn't quite expose the functions needed to
+	// safely parse a sequence of concatenated CIDs from a stream, and we probably don't want to
+	// (re-)implement that here. See https://github.com/ipfs/go-cid/issues/93.
+	// The empty key has value "" (no encoded-zero prefix).
+	value string
+}
+
+// NewTipSetKey builds a new key from a slice of CIDs.
+// The CIDs are assumed to be ordered correctly.
+func NewTipSetKey(cids ...cid.Cid) TipSetKey {
+	encoded, err := encodeKey(cids)
+	if err != nil {
+		panic("failed to encode CIDs: " + err.Error())
+	}
+	return TipSetKey{string(encoded)}
+}
+
+// TipSetKeyFromBytes wraps an encoded key, validating correct decoding.
+func TipSetKeyFromBytes(encoded []byte) (TipSetKey, error) {
+	_, err := decodeKey(encoded)
+	if err != nil {
+		return TipSetKey{}, err
+	}
+	return TipSetKey{string(encoded)}, nil
+}
+
+// Cids returns a slice of the CIDs comprising this key.
+func (k TipSetKey) Cids() []cid.Cid {
```

Comment:
let's hope we don't have to call this much if ever. The allocations are scaring me already

Reply:
I'd say we should cache the cids in the struct, but then we can't use it as a mapkey anymore